avatars/animationHelpers.js

```diff
@@ -412,1069 +423,789 @@ export const loadPromise = (async () => {
-  const idleAnimationOther = _getIdleAnimation('crouch');
-
-  const angleToClosestAnimation = Math.abs(angleDifference(angle, keyWalkAnimationAnglesMirror[0].angle));
-  const angleBetweenAnimations = Math.abs(angleDifference(keyWalkAnimationAnglesMirror[0].angle, keyWalkAnimationAnglesMirror[1].angle));
-  const angleFactor = (angleBetweenAnimations - angleToClosestAnimation) / angleBetweenAnimations;
-  const isBackward = _getAngleToBackwardAnimation(keyWalkAnimationAnglesMirror) < Math.PI * 0.4;
-  if (isBackward !== avatar.lastIsBackward) {
-    avatar.backwardAnimationSpec = {
-      startFactor: avatar.lastBackwardFactor,
-      endFactor: isBackward ? 1 : 0,
-      startTime: now,
-      endTime: now + 150,
     };
-    avatar.lastIsBackward = isBackward;
-  }
-  let mirrorFactor;
-  if (avatar.backwardAnimationSpec) {
-    const f = (now - avatar.backwardAnimationSpec.startTime) / (avatar.backwardAnimationSpec.endTime - avatar.backwardAnimationSpec.startTime);
-    if (f >= 1) {
-      mirrorFactor = avatar.backwardAnimationSpec.endFactor;
-      avatar.backwardAnimationSpec = null;
-    } else {
-      mirrorFactor = avatar.backwardAnimationSpec.startFactor +
-        Math.pow(
-          f,
-          0.5,
-        ) * (avatar.backwardAnimationSpec.endFactor - avatar.backwardAnimationSpec.startFactor);
-    }
-  } else {
-    mirrorFactor = isBackward ? 1 : 0;
-  }
-  avatar.lastBackwardFactor = mirrorFactor;
-
-  if (avatar.emoteAnimation !== avatar.lastEmoteAnimation) {
-    avatar.lastEmoteTime = avatar.emoteAnimation ? now : 0;
   }
-  avatar.lastEmoteAnimation = avatar.emoteAnimation;
-
-  const _getHorizontalBlend = (k, lerpFn, isPosition, target) => {
-    _get7wayBlend(
-      keyWalkAnimationAngles,
-      keyWalkAnimationAnglesMirror,
-      keyRunAnimationAngles,
-      keyRunAnimationAnglesMirror,
-      idleAnimation,
-      // mirrorFactor,
-      // angleFactor,
-      // walkRunFactor,
-      // idleWalkFactor,
-      k,
-      lerpFn,
-      isPosition,
-      localQuaternion,
-    );
-    _get7wayBlend(
-      keyAnimationAnglesOther,
-      keyAnimationAnglesOtherMirror,
-      keyAnimationAnglesOther,
-      keyAnimationAnglesOtherMirror,
-      idleAnimationOther,
-      // mirrorFactor,
-      // angleFactor,
-      // walkRunFactor,
-      // idleWalkFactor,
-      k,
-      lerpFn,
-      isPosition,
-      localQuaternion2,
-    );
-
-    // _get5wayBlend(keyAnimationAnglesOther, keyAnimationAnglesOtherMirror, idleAnimationOther, mirrorFactor, angleFactor, speedFactor, k, lerpFn, localQuaternion2);
 
-    lerpFn
-      .call(
-        target.copy(localQuaternion),
-        localQuaternion2,
-        crouchFactor,
-      );
-  };
-  const _handleDefault = spec => {
-    const {
-      animationTrackName: k,
-      dst,
-      // isTop,
-      lerpFn,
-      isPosition,
-    } = spec;
-
-    _getHorizontalBlend(k, lerpFn, isPosition, dst);
+  // end util ---
+
+  const createMotions = () => {
```

Comment:
Since this setup seems constant for avatars, it does not make much sense to do it in JS. It's just extra overhead, initialization, and GC pressure per avatar. The whole point of doing it in WASM is to remove this overhead from the JS side.

The node system is a great design, but the same thing could be set up more efficiently in C++ instead, and I don't expect many people will be modifying this on the JS side. This is highly entangled with the avatar system anyway.

Reply:
Deleted all motions/nodes on JS side, moved to WASM side.